File: pos_bahrain/api/material_request.py
import frappe
from frappe import _
from frappe.model.mapper import get_mapped_doc
from frappe.utils import flt
from toolz import first


@frappe.whitelist()
def make_stock_entry(source_name, target_doc=None):
    def update_item(obj, target, source_parent):
        qty = (
            flt(flt(obj.stock_qty) - flt(obj.ordered_qty)) / target.conversion_factor
            if flt(obj.stock_qty) > flt(obj.ordered_qty)
            else 0
        )
        target.qty = qty
        target.transfer_qty = qty * obj.conversion_factor
        target.conversion_factor = obj.conversion_factor

        if source_parent.material_request_type == "Material Transfer":
            target.t_warehouse = obj.warehouse
        else:
            target.s_warehouse = obj.warehouse

    def set_missing_values(source, target):
        target.target_branch = _get_branch_by_warehouse(source.pb_to_warehouse)

    doclist = get_mapped_doc(
        "Material Request",
        source_name,
        {
            "Material Request": {
                "doctype": "Stock Transfer",
                "validation": {
                    "docstatus": ["=", 1],
                    "material_request_type": [
                        "in",
                        ["Material Transfer"],
                    ],
                },
            },
            "Material Request Item": {
                "doctype": "Stock Transfer Item",
                "field_map": {
                    "name": "material_request_item",
                    "parent": "material_request",
                    "uom": "stock_uom",
                },
                "postprocess": update_item,
                "condition": lambda doc: doc.ordered_qty < doc.stock_qty,
            },
        },
        target_doc,
        set_missing_values,
    )

    return doclist


def _get_branch_by_warehouse(warehouse):
    data = frappe.db.sql(
        """
        SELECT name FROM `tabBranch`
        WHERE warehouse = %s
        """,
        warehouse,
        as_dict=1
    )
    if not data:
        frappe.throw(_("No branch is associated with Warehouse {}".format(warehouse)))
    return first(data).get("name")

File: pos_bahrain/public/js/material_request.js
frappe.ui.form.on('Material Request', {
  setup: function (frm) {
    frm.set_query('pb_to_warehouse', function () {
      return {
        filters: {
          company: frm.doc.company,
          is_group: 0,
        },
      };
    });
  },
  refresh: function (frm) {
    _make_custom_buttons(frm);
  },
  pb_to_warehouse: function (frm) {
    _set_items_warehouse(frm);
  },
});

function _make_custom_buttons(frm) {
  if (frm.doc.docstatus !== 1) {
    return;
  }
  if (
    frm.doc.material_request_type === 'Material Transfer' &&
    frm.doc.status !== 'Transferred' &&
    frappe.user.has_role('Stock Manager')
  ) {
    frm.add_custom_button(__('Stock Transfer'), () =>
      _make_stock_transfer(frm)
    );
  }
}

function _make_stock_transfer(frm) {
  frappe.model.open_mapped_doc({
    method: 'pos_bahrain.api.material_request.make_stock_entry',
    frm: frm,
  });
}

function _set_items_warehouse(frm) {
  for (const item of frm.doc.items) {
    frappe.model.set_value(
      item.doctype,
      item.name,
      'warehouse',
      frm.doc.pb_to_warehouse
    );
  }
}
